File: rsdroid/src/main/java/net/ankiweb/rsdroid/database/RustSupportSQLiteDatabase.java
/*
 * Copyright (c) 2020 David Allison <davidallisongithub@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 3 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Contains code under the following license
 *
 *     Copyright (C) 2006 The Android Open Source Project
 *     Copyright (C) 2016 The Android Open Source Project
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *          http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 *
 * from android.database.sqlite.SQLiteStatement
 * from update/insert/delete
 */

package net.ankiweb.rsdroid.database;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.SQLException;
import android.database.sqlite.SQLiteTransactionListener;
import android.os.CancellationSignal;
import android.util.Pair;

import androidx.sqlite.db.SupportSQLiteDatabase;
import androidx.sqlite.db.SupportSQLiteQuery;
import androidx.sqlite.db.SupportSQLiteStatement;

import net.ankiweb.rsdroid.BackendV1;

import java.util.List;
import java.util.Locale;
import java.util.Map;

import static android.text.TextUtils.isEmpty;

public class RustSupportSQLiteDatabase implements SupportSQLiteDatabase {
    private static final String[] CONFLICT_VALUES = new String[]
            {"", " OR ROLLBACK ", " OR ABORT ", " OR FAIL ", " OR IGNORE ", " OR REPLACE "};

    private final ThreadLocal<Session> sessionFactory;
    private final boolean isReadOnly;
    private boolean isOpen;

    public RustSupportSQLiteDatabase(BackendV1 backend, boolean readOnly) {
        if (backend == null) {
            throw new IllegalArgumentException("backend was null");
        }
        this.sessionFactory = new SessionThreadLocal(backend);
        this.isReadOnly = readOnly;
        this.isOpen = true;
    }

    @Override
    public boolean isReadOnly() {
        return isReadOnly;
    }

    @Override
    public boolean isOpen() {
        return isOpen;
    }

    @Override
    public SupportSQLiteStatement compileStatement(String sql) {
        return new RustSQLiteStatement(this, sql);
    }

    @Override
    public void beginTransaction() {
        getSession().beginTransaction();
    }

    @Override
    public void endTransaction() {
        getSession().endTransaction();
    }

    @Override
    public void setTransactionSuccessful() {
        getSession().setTransactionSuccessful();
    }

    @Override
    public boolean inTransaction() {
        return getSession().inTransaction();
    }

    @Override
    public int getVersion() {
        throw NotImplementedException.todo();
    }

    @Override
    public void setVersion(int version) {
        throw NotImplementedException.todo();
    }

    @Override
    public Cursor query(String query) {
        return query(query, null);
    }

    @Override
    public Cursor query(String query, Object[] bindArgs) {
        return new StreamingProtobufSQLiteCursor(getSession(), query, bindArgs);
    }


    @Override
    public Cursor query(SupportSQLiteQuery query) {
        throw NotImplementedException.todo();
    }

    @Override
    public Cursor query(SupportSQLiteQuery query, CancellationSignal cancellationSignal) {
        throw NotImplementedException.todo();
    }

    @Override
    public long insert(String table, int conflictAlgorithm, ContentValues values) throws SQLException {
        StringBuilder sql = new StringBuilder();
        sql.append("INSERT");
        sql.append(CONFLICT_VALUES[conflictAlgorithm]);
        sql.append(" INTO ");
        sql.append(table);
        sql.append('(');

        Object[] bindArgs = null;
        int size = (values != null && values.size() > 0)
                ? values.size() : 0;
        if (size > 0) {
            bindArgs = new Object[size];
            int i = 0;
            for (Map.Entry<String, Object> entry : values.valueSet()) {
                sql.append((i > 0) ? "," : "");
                sql.append(entry.getKey());
                bindArgs[i++] = entry.getValue();
            }
            sql.append(')');
            sql.append(" VALUES (");
            for (i = 0; i < size; i++) {
                sql.append((i > 0) ? ",?" : "?");
            }
        } else {
            sql.append((String) null).append(") VALUES (NULL");
        }
        sql.append(')');

        query(sql.toString(), bindArgs).close();
        return 0;
    }

    @Override
    public int update(String table, int conflictAlgorithm, ContentValues values, String whereClause, Object[] whereArgs) {
        // taken from SQLiteDatabase class.
        if (values == null || values.size() == 0) {
            throw new IllegalArgumentException("Empty values");
        }
        StringBuilder sql = new StringBuilder(120);
        sql.append("UPDATE ");
        sql.append(CONFLICT_VALUES[conflictAlgorithm]);
        sql.append(table);
        sql.append(" SET ");

        // move all bind args to one array
        int setValuesSize = values.size();
        int bindArgsSize = (whereArgs == null) ? setValuesSize : (setValuesSize + whereArgs.length);
        Object[] bindArgs = new Object[bindArgsSize];
        int i = 0;
        for (String colName : values.keySet()) {
            sql.append((i > 0) ? "," : "");
            sql.append(colName);
            bindArgs[i++] = values.get(colName);
            sql.append("=?");
        }
        if (whereArgs != null) {
            for (i = setValuesSize; i < bindArgsSize; i++) {
                bindArgs[i] = whereArgs[i - setValuesSize];
            }
        }
        if (!isEmpty(whereClause)) {
            sql.append(" WHERE ");
            sql.append(whereClause);
        }

        return this.executeGetRowsAffected(sql.toString(), bindArgs);
    }

    @Override
    public void execSQL(String sql) throws SQLException {
        execSQL(sql, null);
    }

    @Override
    public void execSQL(String sql, Object[] bindArgs) throws SQLException {
        query(sql, bindArgs).close();
    }

    @Override
    public boolean needUpgrade(int newVersion) {
        // needed for metaDB, but not for Anki DB
        throw NotImplementedException.todo();
    }

    @Override
    public String getPath() {
        return getSession().getPath();
    }


    @Override
    public boolean isWriteAheadLoggingEnabled() {
        return false;
    }

    @Override
    public void disableWriteAheadLogging() {
        // Nothing to do - openAnkiDroidCollection does this
    }

    @Override
    public boolean isDatabaseIntegrityOk() {
        Cursor pragma_integrity_check = query("pragma integrity_check");
        if (!pragma_integrity_check.moveToFirst()) {
            return false;
        }
        String value = pragma_integrity_check.getString(0);
        return "ok".equals(value);
    }

    @Override
    public void close() {
        isOpen = false;
        getSession().closeDatabase();
    }

    /* Not part of interface */

    public int executeGetRowsAffected(String sql, Object[] bindArgs) {
        try {
            return getSession().executeGetRowsAffected(sql, bindArgs);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public long insertForForId(String sql, Object[] bindArgs) {
        try {
            return getSession().insertForId(sql, bindArgs);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    /** Helper methods */

    private Session getSession() {
        return sessionFactory.get();
    }

    /** Confirmed that the below are not used for our code */


    @Override
    public int delete(String table, String whereClause, Object[] whereArgs) {
        // Complex to implement and not required
        throw new NotImplementedException();
    }

    @Override
    public boolean isDbLockedByCurrentThread() {
        throw new NotImplementedException();
    }

    @Override
    public boolean yieldIfContendedSafely() {
        throw new NotImplementedException();
    }

    @Override
    public boolean yieldIfContendedSafely(long sleepAfterYieldDelay) {
        throw new NotImplementedException();
    }

    @Override
    public void setLocale(Locale locale) {
        throw new NotImplementedException();
    }

    @Override
    public void setMaxSqlCacheSize(int cacheSize) {
        throw new NotImplementedException();
    }

    @Override
    public long getMaximumSize() {
        throw new NotImplementedException();
    }

    @Override
    public long setMaximumSize(long numBytes) {
        throw new NotImplementedException();
    }

    @Override
    public long getPageSize() {
        throw new NotImplementedException();
    }

    @Override
    public void setPageSize(long numBytes) {
        throw new NotImplementedException();
    }


    @Override
    public void setForeignKeyConstraintsEnabled(boolean enable) {
        throw new NotImplementedException();
    }


    @Override
    public boolean enableWriteAheadLogging() {
        throw new NotImplementedException();
    }

    @Override
    public List<Pair<String, String>> getAttachedDbs() {
        throw new NotImplementedException();
    }

    @Override
    public void beginTransactionNonExclusive() {
        throw new NotImplementedException();
    }

    @Override
    public void beginTransactionWithListener(SQLiteTransactionListener transactionListener) {
        throw new NotImplementedException();
    }

    @Override
    public void beginTransactionWithListenerNonExclusive(SQLiteTransactionListener transactionListener) {
        throw new NotImplementedException();
    }
}

File: rsdroid/src/main/java/net/ankiweb/rsdroid/database/RustSupportSQLiteDatabase.kt
/*
 * Copyright (c) 2020 David Allison <davidallisongithub@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 3 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Contains code under the following license
 *
 *     Copyright (C) 2006 The Android Open Source Project
 *     Copyright (C) 2016 The Android Open Source Project
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *          http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 *
 * from android.database.sqlite.SQLiteStatement
 * from update/insert/delete
 */
package net.ankiweb.rsdroid.database

import android.content.ContentValues
import android.database.Cursor
import android.database.SQLException
import android.database.sqlite.SQLiteTransactionListener
import android.os.CancellationSignal
import android.text.TextUtils
import android.util.Pair
import androidx.sqlite.db.SupportSQLiteDatabase
import androidx.sqlite.db.SupportSQLiteQuery
import androidx.sqlite.db.SupportSQLiteStatement
import net.ankiweb.rsdroid.BackendV1
import java.util.*

class RustSupportSQLiteDatabase(backend: BackendV1?, readOnly: Boolean) : SupportSQLiteDatabase {
    private val sessionFactory: ThreadLocal<Session>
    private val isReadOnly: Boolean
    private var isOpen: Boolean
    override fun isReadOnly(): Boolean {
        return isReadOnly
    }

    override fun isOpen(): Boolean {
        return isOpen
    }

    override fun compileStatement(sql: String): SupportSQLiteStatement {
        return RustSQLiteStatement(this, sql)
    }

    override fun beginTransaction() {
        session.beginTransaction()
    }

    override fun endTransaction() {
        session.endTransaction()
    }

    override fun setTransactionSuccessful() {
        session.setTransactionSuccessful()
    }

    override fun inTransaction(): Boolean {
        return session.inTransaction()
    }

    override fun getVersion(): Int {
        throw NotImplementedException.Companion.todo()
    }

    override fun setVersion(version: Int) {
        throw NotImplementedException.Companion.todo()
    }

    override fun query(query: String): Cursor {
        return query(query, null)
    }

    override fun query(query: String, bindArgs: Array<Any?>?): Cursor {
        return StreamingProtobufSQLiteCursor(session, query, bindArgs)
    }

    override fun query(query: SupportSQLiteQuery): Cursor {
        throw NotImplementedException.Companion.todo()
    }

    override fun query(query: SupportSQLiteQuery, cancellationSignal: CancellationSignal): Cursor {
        throw NotImplementedException.Companion.todo()
    }

    @Throws(SQLException::class)
    override fun insert(table: String, conflictAlgorithm: Int, values: ContentValues): Long {
        val sql = StringBuilder()
        sql.append("INSERT")
        sql.append(CONFLICT_VALUES[conflictAlgorithm])
        sql.append(" INTO ")
        sql.append(table)
        sql.append('(')
        var bindArgs: Array<Any?>? = null
        val size = if (values != null && values.size() > 0) values.size() else 0
        if (size > 0) {
            bindArgs = arrayOfNulls(size)
            var i = 0
            for ((key, value) in values.valueSet()) {
                sql.append(if (i > 0) "," else "")
                sql.append(key)
                bindArgs[i++] = value
            }
            sql.append(')')
            sql.append(" VALUES (")
            i = 0
            while (i < size) {
                sql.append(if (i > 0) ",?" else "?")
                i++
            }
        } else {
            sql.append(null as String?).append(") VALUES (NULL")
        }
        sql.append(')')
        query(sql.toString(), bindArgs).close()
        return 0
    }

    override fun update(table: String, conflictAlgorithm: Int, values: ContentValues, whereClause: String, whereArgs: Array<Any?>?): Int {
        // taken from SQLiteDatabase class.
        require(!(values == null || values.size() == 0)) { "Empty values" }
        val sql = StringBuilder(120)
        sql.append("UPDATE ")
        sql.append(CONFLICT_VALUES[conflictAlgorithm])
        sql.append(table)
        sql.append(" SET ")

        // move all bind args to one array
        val setValuesSize = values.size()
        val bindArgsSize = if (whereArgs == null) setValuesSize else setValuesSize + whereArgs.size
        val bindArgs = arrayOfNulls<Any>(bindArgsSize)
        var i = 0
        for (colName in values.keySet()) {
            sql.append(if (i > 0) "," else "")
            sql.append(colName)
            bindArgs[i++] = values[colName]
            sql.append("=?")
        }
        if (whereArgs != null) {
            i = setValuesSize
            while (i < bindArgsSize) {
                bindArgs[i] = whereArgs[i - setValuesSize]
                i++
            }
        }
        if (!TextUtils.isEmpty(whereClause)) {
            sql.append(" WHERE ")
            sql.append(whereClause)
        }
        return executeGetRowsAffected(sql.toString(), bindArgs)
    }

    @Throws(SQLException::class)
    override fun execSQL(sql: String) {
        execSQL(sql, null)
    }

    @Throws(SQLException::class)
    override fun execSQL(sql: String, bindArgs: Array<Any?>?) {
        query(sql, bindArgs).close()
    }

    override fun needUpgrade(newVersion: Int): Boolean {
        // needed for metaDB, but not for Anki DB
        throw NotImplementedException.Companion.todo()
    }

    override fun getPath(): String {
        return session.getPath()
    }

    override fun isWriteAheadLoggingEnabled(): Boolean {
        return false
    }

    override fun disableWriteAheadLogging() {
        // Nothing to do - openAnkiDroidCollection does this
    }

    override fun isDatabaseIntegrityOk(): Boolean {
        val pragma_integrity_check = query("pragma integrity_check")
        if (!pragma_integrity_check.moveToFirst()) {
            return false
        }
        val value = pragma_integrity_check.getString(0)
        return "ok" == value
    }

    override fun close() {
        isOpen = false
        session.closeDatabase()
    }

    /* Not part of interface */
    fun executeGetRowsAffected(sql: String, bindArgs: Array<Any?>?): Int {
        return try {
            session.executeGetRowsAffected(sql, bindArgs)
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
    }

    fun insertForForId(sql: String, bindArgs: Array<Any?>?): Long {
        return try {
            session.insertForId(sql, bindArgs)
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
    }

    /** Helper methods  */
    private val session: Session
        private get() = sessionFactory.get()

    /** Confirmed that the below are not used for our code  */
    override fun delete(table: String, whereClause: String, whereArgs: Array<Any>): Int {
        // Complex to implement and not required
        throw NotImplementedException()
    }

    override fun isDbLockedByCurrentThread(): Boolean {
        throw NotImplementedException()
    }

    override fun yieldIfContendedSafely(): Boolean {
        throw NotImplementedException()
    }

    override fun yieldIfContendedSafely(sleepAfterYieldDelay: Long): Boolean {
        throw NotImplementedException()
    }

    override fun setLocale(locale: Locale) {
        throw NotImplementedException()
    }

    override fun setMaxSqlCacheSize(cacheSize: Int) {
        throw NotImplementedException()
    }

    override fun getMaximumSize(): Long {
        throw NotImplementedException()
    }

    override fun setMaximumSize(numBytes: Long): Long {
        throw NotImplementedException()
    }

    override fun getPageSize(): Long {
        throw NotImplementedException()
    }

    override fun setPageSize(numBytes: Long) {
        throw NotImplementedException()
    }

    override fun setForeignKeyConstraintsEnabled(enable: Boolean) {
        throw NotImplementedException()
    }

    override fun enableWriteAheadLogging(): Boolean {
        throw NotImplementedException()
    }

    override fun getAttachedDbs(): List<Pair<String, String>> {
        throw NotImplementedException()
    }

    override fun beginTransactionNonExclusive() {
        throw NotImplementedException()
    }

    override fun beginTransactionWithListener(transactionListener: SQLiteTransactionListener) {
        throw NotImplementedException()
    }

    override fun beginTransactionWithListenerNonExclusive(transactionListener: SQLiteTransactionListener) {
        throw NotImplementedException()
    }

    companion object {
        private val CONFLICT_VALUES = arrayOf("", " OR ROLLBACK ", " OR ABORT ", " OR FAIL ", " OR IGNORE ", " OR REPLACE ")
    }

    init {
        requireNotNull(backend) { "backend was null" }
        sessionFactory = SessionThreadLocal(backend)
        isReadOnly = readOnly
        isOpen = true
    }
}